README
libehnet64: Easy Host-Network 64
------------------------------
multi-platfrom 64-bit versions of functions to convert between network and host byte orders


hton64 and ntoh64
-----------------
64-bit function to convert between network and host byte order

The header arpa/inet.h declares:

       uint32_t htonl(uint32_t host32);
       uint32_t ntohl(uint32_t net32);

       uint16_t htons(uint16_t host16);
       uint16_t ntohs(uint16_t net16);

however, nothing for 64 bit. Depending upon the platform, various
headers exist which implement similar 64 bit functions, but nothing as
standard as arpa/inet.h. The header ehnet64.h aims to work on any
platform and #defines these:

       uint64_t hton64(uint64_t host64);
       uint64_t ntoh64(uint64_t net64);


(I elected *not* to name the functions "htonll" and "ntohll" because
"long" and "long long" are the same on 64bit arch, and I think the 64 is
less ambiguous. You may wish to '#define htonll hton64' and '#define
ntohll ntoh64' in your code.)


Usage
-----

If using Autoconf, consider adding these to your configure.ac:

AC_CHECK_HEADERS([byteswap.h endian.h inttypes.h stdint.h])
AC_CHECK_FUNCS([bswap_32 bswap_64])

If defining by hand, define the applicable of the following:

#define HAVE_INTTYPES_H 1
#define HAVE_STDINT_H 1
#define HAVE_ENDIAN_H 1
#define HAVE_BYTESWAP_H 1
#define HAVE_BSWAP_32 1 /* not #defined actually a function */
#define HAVE_BSWAP_64 1 /* not #defined actually a function */

Finally, #include "ehnet64.h" into your source in order to make use of
the functions "hton64" and "ntoh64" in your program.


Building and Testing
--------------------
autoreconf -iv &&
 ./configure &&
 make check &&
 echo "Success."


Packaging
---------
# and if you wish:
make distcheck


License
-------
GNU Lesser General Public License (LGPL), version 3 or later.
See COPYING and COPYING.GPL3 for details.


ChangeLog
2017-06-24  Eric Herman <eric@freesa.org>

	Initial version

	* src/ehnet64.h: #define ntoh64(x) and hton64(x) macros
	* tests/test-hton64.c: one simple hton64(x) test
	* tests/test-ntoh64.c: one simple ntoh64(x) test


tests/test-hton64.c
/* test-hton64.c
Copyright (C) 2017 Eric Herman <eric@freesa.org>

This work is free software: you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This work is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
License for more details.
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include "src/ehnet64.h"

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

union endian_test_u {
	uint8_t bytes[8];
	uint64_t n;
};

int main(int argc, char **argv)
{
	union endian_test_u litend, bigend;
	uint64_t val, hostv, result;
	size_t i;
	int verbose;
	char *end;

	verbose = (argc > 1) ? atoi(argv[1]) : 0;
	val = (argc > 2) ? strtoull(argv[2], &end, 10) : 0x1122334455667788ULL;

	if (verbose) {
		printf("value  = %016llx (%llu)\n", (unsigned long long)val,
		       (unsigned long long)val);
	}

	for (i = 0; i < 8; ++i) {
		litend.bytes[i] = ((val & (0xFFULL << (8 * i))) >> (8 * i));
		bigend.bytes[7 - i] = litend.bytes[i];
	}

	if (verbose) {
		printf("litend = %016llx (%llu)\n",
		       (unsigned long long)litend.n,
		       (unsigned long long)litend.n);
	}
	if (verbose) {
		printf("bigend = %016llx (%llu)\n",
		       (unsigned long long)bigend.n,
		       (unsigned long long)bigend.n);
	}

	hostv = (bigend.n == val) ? bigend.n : litend.n;

	if (verbose) {
		printf("host v = %016llx (%llu)\n", (unsigned long long)hostv,
		       (unsigned long long)hostv);
	}

	result = hton64(hostv);
	if (verbose) {
		printf("result = hton64(%016llx)\n", (unsigned long long)hostv);
		printf("result = %016llx (%llu)\n", (unsigned long long)result,
		       (unsigned long long)result);
	}

	if (result != bigend.n) {
		fprintf(stderr, "%0llx = hton64(%016llx), expected: %016llx\n",
			(unsigned long long)result, (unsigned long long)hostv,
			(unsigned long long)bigend.n);
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}


tests/test-ntoh64.c
/* test-ntoh64.c
Copyright (C) 2017 Eric Herman <eric@freesa.org>

This work is free software: you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This work is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
License for more details.
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include "src/ehnet64.h"

#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

union endian_test_u {
	uint8_t bytes[8];
	uint64_t n;
};

int main(int argc, char **argv)
{
	union endian_test_u litend, bigend;
	uint64_t val, hostv, result;
	size_t i;
	int verbose;
	char *end;

	verbose = (argc > 1) ? atoi(argv[1]) : 0;
	val = (argc > 2) ? strtoull(argv[2], &end, 10) : 0x1122334455667788ULL;

	if (verbose) {
		printf("value  = %016llx (%llu)\n", (unsigned long long)val,
		       (unsigned long long)val);
	}

	for (i = 0; i < 8; ++i) {
		litend.bytes[i] = ((val & (0xFFULL << (8 * i))) >> (8 * i));
		bigend.bytes[7 - i] = litend.bytes[i];
	}

	if (verbose) {
		printf("litend = %016llx (%llu)\n",
		       (unsigned long long)litend.n,
		       (unsigned long long)litend.n);
	}
	if (verbose) {
		printf("bigend = %016llx (%llu)\n",
		       (unsigned long long)bigend.n,
		       (unsigned long long)bigend.n);
	}

	hostv = (bigend.n == val) ? bigend.n : litend.n;

	if (verbose) {
		printf("host v = %016llx (%llu)\n", (unsigned long long)hostv,
		       (unsigned long long)hostv);
	}

	result = ntoh64(bigend.n);
	if (verbose) {
		printf("result = ntoh64(%016llx)\n",
		       (unsigned long long)bigend.n);
		printf("result = %016llx (%llu)\n", (unsigned long long)result,
		       (unsigned long long)result);
	}

	if (result != hostv) {
		fprintf(stderr, "%0llx = ntoh64(%016llx), expected: %016llx\n",
			(unsigned long long)result,
			(unsigned long long)bigend.n,
			(unsigned long long)hostv);
		return EXIT_FAILURE;
	}

	return EXIT_SUCCESS;
}


src/ehnet64.h
/* ehnet64.h 64 versions of ntohl and htonl as ntoh64 and hton64
Copyright (C) 2017 Eric Herman <eric@freesa.org>

This work is free software: you can redistribute it and/or modify it
under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or (at
your option) any later version.

This work is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
License for more details.
*/

/*
The header arpa/inet.h declares:

       uint32_t htonl(uint32_t host32);
       uint32_t ntohl(uint32_t net32);

       uint16_t htons(uint16_t host16);
       uint16_t ntohs(uint16_t net16);

however, nothing for 64 bit ... below we #define these:

       uint64_t hton64(uint64_t host64);
       uint64_t ntoh64(uint64_t net64);
*/

#ifndef EHNET64_H
#define EHNET64_H

#ifdef __cplusplus
extern "C" {
#endif

#if HAVE_ENDIAN_H
#include <endian.h>
#define hton64(x) htobe64(x)
#define ntoh64(x) be64toh(x)
#else /* HAVE_ENDIAN_H */

#if HAVE_BYTESWAP_H
#include <byteswap.h>
#endif /* HAVE_BYTESWAP_H */

#if (defined(bswap_64) || HAVE_BSWAP_64)
#define ehnet64_bswap_64(x) bswap_64(x)
#endif /* ifdef bswap_64 */

#if (defined(bswap_32) || HAVE_BSWAP_32)
#define ehnet64_bswap_32(x) bswap_32(x)
#endif /* ifdef bswap_32 */

#ifndef ehnet64_bswap_32
/* okay fine, we'll roll our own by hand */
#define ehnet64_bswap_32(x) \
	((((x) & 0xff000000) >> 24) | (((x) & 0x00ff0000) >>  8) | \
	 (((x) & 0x0000ff00) <<  8) | (((x) & 0x000000ff) << 24))
#endif /* ehnet64_bswap_32 */

#if HAVE_STDINT_H
#include <inttypes.h>		/* uint64_t uint32_t */
#endif /* HAVE_INTTYPES_H */

#if HAVE_STDINT_H
#include <stdint.h>		/* uint64_t uint32_t */
#endif /* HAVE_STDINT_H */

#ifndef ehnet64_bswap_64
#define ehnet64_bswap_64(x) \
	((((uint64_t)ehnet64_bswap_32((uint32_t)x) << 32) | \
          ((uint64_t)ehnet64_bswap_32((uint32_t)(x >> 32)))))
#endif /* ehnet64_bswap_64 */

#define ehnet64_bigendian_host (ehnet64_bswap_32(1U) == 1U)

#define hton64(x) (ehnet64_bigendian_host ? (uint64_t)x : ehnet64_bswap_64(x))

#define ntoh64(x) (ehnet64_bigendian_host ? (uint64_t)x : ehnet64_bswap_64(x))

#endif /* HAVE_ENDIAN_H */

#ifdef __cplusplus
}
#endif

#endif /* EHNET64_H */
